### PR TITLE
GraphLinks: replace trait with enum

### PR DESCRIPTION
**Pull request by xzfc**
~Depends on #5666 and #5492. Read only the last commit during the review.~

This PR improves the performance of graph links by merging `impl GraphLinks`, `struct GraphLinksRam` and `struct GraphLinksMmap` into a single `struct GraphLinks`.

The new structure is self-referential (using [`self_cell`](https://docs.rs/self_cell/1.1.0) crate), and contains `GraphLinksView` with precalculated slices.
Enum dispatching is not required: once `GraphLinksView` is created during loading, it doesn't care about the underlying storage type.

At a high level, the new structure looks like this:
```rust
self_cell::self_cell! {
    pub struct GraphLinks {
        owner: GraphLinksEnum,     // owns the data
        #[covariant]
        dependent: GraphLinksView, // slice references to the data
    }
}

enum GraphLinksEnum {
    Ram(Vec<u8>),
    Mmap(Arc<Mmap>),
}

struct GraphLinksView<'a> {
    reindex: &'a [PointOffsetType],
    offsets: &'a [u64],
    compression: CompressionInfo<'a>,
    level_offsets: Vec<u64>,
}

enum CompressionInfo<'a> { … }
```

The new implementation is more efficient, because slices are created only once during loading and not recalculated on every access, so we can avoid additional indirections and bounds checks. For comparison:
```rust
// old
let offsets_range: Range<usize> = self.info.offsets_range();
let offsets: &[u64] = <[u64]>::ref_from_bytes(&self.data[offsets_range]).unwrap(); // indirection + bounds check
let links_start: Range<usize> = offsets[idx] as usize..offsets[idx + 1] as usize;

// new
let links_range: Range<usize> = self.offsets[idx] as usize..self.offsets[idx + 1] as usize;
```

On the test introduced in #5666, the new implementation makes the HNSW search 2-3% faster; and makes compressed version faster than the uncompressed one. I'll provide more benchmark results later.

Besides the performance improvements, replacing a trait with a enum made it possible to simplify some code.

### CODE REVIEW

File: lib/segment/src/index/vector_index_base.rs
Comment (IvanPleshkov) [inline lib/segment/src/index/vector_index_base.rs:60]: Honestly, it looks inconsistent to me. Each variant has `Ram`/`Mmap`/`Compressed` versions and only HNSW index is an exception with internal definitions. Like quantization and vector storage, we do the same: expand all possible options in a single enum. Future developers may be confused and think that hnsw does not have ram/mmap versions. I won't say it's a blocker, but let's discuss it. Probably we can inline all `HNSWIndex` variants here

File: lib/segment/benches/hnsw_search_graph.rs
Comment (IvanPleshkov) [inline lib/segment/benches/hnsw_search_graph.rs:35]: Are this changes from https://github.com/qdrant/qdrant/pull/5666? Did you add something new or is it copy-paste from another PR?

**Comment by generall** [comment]
@xzfc could you please resolve merge conflict?

File: lib/segment/benches/hnsw_search_graph.rs
Comment (xzfc) [reply lib/segment/benches/hnsw_search_graph.rs:35]: Yes, this PR is based on #5666. Only the last commit is related to this PR. Check the commit tree.

File: lib/segment/src/index/vector_index_base.rs
Comment (xzfc) [reply lib/segment/src/index/vector_index_base.rs:60]: TBH I don't see an issue here. Sparse is already a different beast than HNSW/dense, and now they are different just a bit more. Whether an enum is expanded here I see an implementation detail of each index.

I did two things here:
- Removed a generic parameter from `HNSWIndex`. Pros: lesser compiler times and binary sizes.
- Removed an enum variant from `VectorIndexEnum` as it duplicates an internal runtime flag of `HNSWIndex`. Among other things, this change makes an [Illegal State Unrepresentable](https://corrode.dev/blog/illegal-state/), i.e. we can't have `on_disk: true` with `HnswRam` by accident.

Comment (IvanPleshkov) [reply lib/segment/src/index/vector_index_base.rs:60]: @timvisee needs additional opinion. After a personal discussion with @xzfc , we don't have a strong vision in the same direction

Comment (coszio) [reply lib/segment/src/index/vector_index_base.rs:63]: My personal unsolicited take: @xzfc makes good arguments, for me it makes sense to go with this new design, which partitions vector indices by nature, then by implementation. 

Regarding consistency, I would argue that all the other `Sparse*` variants should undergo a similar refactoring (joining them into a separate enum)

**Review by coszio — COMMENTED** [review]
I like the refactoring into an enum 👍

However, I am failing to see the benefit of the `self_cell` macro. Why can't it just be a `try_new` implementation? I see two cons of using it here:
- it is only used as a pattern rather than a tool for maintaining less code.
- Additional complexity: I did not know about the crate before, and had to go into its docs to understand what it was about.

File: lib/segment/src/index/vector_index_base.rs
Comment (IvanPleshkov) [reply lib/segment/src/index/vector_index_base.rs:63]: I agree that @xzfc proposal is good. But I believe we won't do the same for all variants and all enums with old design

Comment (IvanPleshkov) [reply lib/segment/src/index/vector_index_base.rs:63]: Anyway thanks @coszio! I find your comment as "looks good", resolve this thread

File: lib/segment/src/index/hnsw_index/graph_links.rs
Comment (timvisee) [inline lib/segment/src/index/hnsw_index/graph_links.rs:516]: Can we potentially preallocate this vec?

File: lib/segment/src/index/hnsw_index/graph_links.rs
Comment (timvisee) [inline lib/segment/src/index/hnsw_index/graph_links.rs:523]:
```suggestion
        let mut edges = Vec::with_capacity(self.num_points());
```

File: lib/segment/src/index/vector_index_base.rs
Comment (timvisee) [reply lib/segment/src/index/vector_index_base.rs:63]: Thanks for bringing this to our attention Ivan, it's good to discuss changes with this significance.

I agree that it would be nice to have the same architecture in multiple places, but in my opinion xzfc gives two compelling arguments. Based on that I also vote to keep the design in this PR.

**Review by timvisee — COMMENTED** [review]
On a first review pass I see no problems in this PR.

I'd like to do a second review pass tomorrow to carefully digest all the small details in this PR, after which I'll leave my stamp of approval.

**Comment by coszio** [comment]
> It is to allow self reference without implementing this from scratch.

Yeah, no, after looking deeper into it, I realize why it is needed. It is not possible to _just_ create a struct which references itself. Sorry for the noise!

**Review by coszio — APPROVED** [review]

**Review by IvanPleshkov — APPROVED** [review]

**Review by timvisee — APPROVED** [review]
Thank you for your work on this! :pray:

**Pull request merged**
